### PR TITLE
Improve support for azure cloud provider

### PR DESCRIPTION
**Pull request by abn**
This change set covers the following aspects when using the azure cloud provider.

* Populate controller and api server args facts for azure
* Create `azure.conf` and introduce related configuration
* Enable management storage class configuration to use azure storage
* Configure default storage class along with default configuration

### CODE REVIEW

**Comment by abn** [comment]
@kwoodson can we please get this tested?

**Comment by abn** [comment]
Looks like the test failure is unrelated to the change set. I have resolved the conflict and rebased the branch.

File: roles/openshift_management/tasks/storage/storage.yml [hunk 1]
```diff
@@ -1,3 +1,3 @@
 ---
 - include_tasks: nfs.yml
-  when: not (openshift_cloudprovider_kind is defined and (openshift_cloudprovider_kind == 'aws' or openshift_cloudprovider_kind == 'gce'))
+  when: not (openshift_cloudprovider_kind is defined and (openshift_cloudprovider_kind in ['aws', 'azure', 'gce']))
```
Comment (michaelgugino) [inline roles/openshift_management/tasks/storage/storage.yml]: Let's change this to

```yaml
when: not (openshift_cloudprovider_kind | default('none') in ['aws', 'azure', 'gce'])

File: roles/openshift_sanitize_inventory/tasks/main.yml [hunk 2]
```diff
@@ -77,6 +77,23 @@
...
+        msg: "'Required variable for azure provider '{{ item }}' is not defined"
+      when: vars[item] is not defined
+      with_items:
+        - openshift_cloudprovider_azure_client_id
+        - openshift_cloudprovider_azure_client_secret
+        - openshift_cloudprovider_azure_tenant_id
+        - openshift_cloudprovider_azure_subscription_id
+        - openshift_cloudprovider_azure_resource_group
+        - openshift_cloudprovider_azure_location
+  when:
```
Comment (michaelgugino) [inline roles/openshift_sanitize_inventory/tasks/main.yml]: We like to put the when conditions at the start of the block, much easier to read that way.

File: roles/openshift_sanitize_inventory/tasks/main.yml [hunk 3]
```diff
@@ -77,6 +77,23 @@
         openshift_cloudprovider_vsphere_host is undefined or openshift_cloudprovider_vsphere_datacenter is undefined or
         openshift_cloudprovider_vsphere_datastore is undefined )
 
+- name: Ensure azure configuration variables are defined
+  block:
+    - name: ensure provider configuration variables are defined
+      fail:
+        msg: "'Required variable for azure provider '{{ item }}' is not defined"
+      when: vars[item] is not defined
+      with_items:
```
Comment (michaelgugino) [inline roles/openshift_sanitize_inventory/tasks/main.yml]: No need for with_items here.  Better to check that all are defined at once and provide link to docs with required vars.

```yaml
when: >
  openshift_cloudprovider_azure_client_id is not defined
  or ... is not defined
  or ..
```

Comment (abn) [reply roles/openshift_sanitize_inventory/tasks/main.yml]: Ah. Can change that to recommendation. Just used with items as it'll provide a better error message in this case either all items that failed.

**Comment by abn** [comment]
@michaelgugino updated the commits to include requested changes. I'll send a PR to openshift-docs once this is merged so that the variables are detailed in the documentation linked from the failure message.

**Pull request merged**